### ISSUE TITLE
Benchmarking GPU game frames

### ISSUE THREAD

**Issue by Jasper-Bekkers** [bheisler/criterion.rs#327]
Hi,

I've started experimenting a little bit with using `criterion.rs` for profiling GPU workloads and in the simple cases this works exceedingly well :-)

However, I have a bit more of a complex workload that I'd like to express in your library as well and I can't quite find a way of doing this at the moment and this may or may not be something you can help me with.

Right now I'm benchmarking an entire workload (multiple compute kernels in order) using `iter_custom` kind of like this:

```rust
    criterion.bench_function(
        "Dual contour",
        |b| b.iter_custom(|iters| {
            let mut total = std::time::Duration::new(0, 0);

            for _ in 0..iters {
                let mut cmd = queue.create_command_buffer();
                cmd.begin();

                cmd.reset_query_pool(&qp);

                cmd.write_timestamp(&qp, 0);
                cmd.dispatch(&cs_pipeline, &[&cbuf], 10, 10, 10);
                cmd.dispatch(&mesh_gen_pipeline, &[&mesh_gen_ds], 10, 10, 10);

                cmd.write_timestamp(&qp, 1);

                cmd.resolve_query_pool(&qp);

                cmd.end();

                queue.submit_wait_idle(cmd);

                let times = qp.load(&queue);
                total += times[1] - times[0]
            }
            total
        })
    );
```

This works quite well, but this only gives me the total time for the two calls to `cmd.dispatch` and ideally I'd measure them individually. One way to do this would be of course to create several benchmarks just like above and split it up where the `dispatch` calls happen. This would of course work, but I'm slightly more interested in expressing it another way.

I'd like to be able to run a full game frame, several times (as predicted by criterion), record timestamps between each dispatch call myself and report all of them back to criterion in one go.

```rust
    criterion.bench_function(
        "Dual contour",
        |b| b.iter_custom(|iters| {
            let mut total_cs = std::time::Duration::new(0, 0);
            let mut total_mesh_gen = std::time::Duration::new(0, 0);

            for _ in 0..iters {
                let mut cmd = queue.create_command_buffer();
                cmd.begin();

                cmd.reset_query_pool(&qp);

                cmd.write_timestamp(&qp, 0);
                cmd.dispatch(&cs_pipeline, &[&cbuf], 10, 10, 10);
                cmd.write_timestamp(&qp, 1);
                cmd.dispatch(&mesh_gen_pipeline, &[&mesh_gen_ds], 10, 10, 10);
                cmd.write_timestamp(&qp, 2);

                cmd.resolve_query_pool(&qp);

                cmd.end();

                queue.submit_wait_idle(cmd);

                let times = qp.load(&queue);
                total_cs += times[1] - times[0];
                total_mesh_gen += times[2] - times[1];
            }
            [total_cs, total_mesh_gen] // maybe also return names here for the different timers
        })
    );
```

This has some major advantages:
  - There is less total overhead (submitting command buffers and setting up dispatches and draw calls takes time). 
  - Sometimes, like in this example, the first dispatch will generate work for consecutive dispatches that otherwise I'd need to cache off to intermediate buffers. Since it's on the GPU this work will potentially be generated out-of-order and vary from run to run. Replaying on cached data wouldn't show the potential effect that this has on the execution time since it would most likely reduce the standard deviation of the exec time compared to real-life.
  - It's closer to how in-game profiling is currently done, I'd potentially just have to run my workload several times.

Is there a way to achieve this? Keep in mind that everything recorded to `cmd` will be executed as one consecutive workload that only gets run when `submit_wait_idle` gets called. Only after that function call is it safe to do the `qp.load` call.

**Comment by bheisler** [comment]
Hey, thanks for trying Criterion.rs! Thanks also for your patience, I've been super busy IRL lately and haven't been able to keep up with Criterion.rs as much as I'd like.

Interesting suggestion! I haven't run into a situation like this before. Criterion.rs doesn't support anything like this at the moment. Unfortunately, it would take some internal refactoring to enable supporting something like this, I think. That's not to say that it won't happen, but it would take some work. Unfortunately, due to the aforementioned busyness IRL I probably won't be able to get to it soon.

**Comment by Jasper-Bekkers** [comment]
Closing due to lack of traction.
